File: ScheduleView/Database/WPF_VB/Models/SqlAppointment.vb
Imports Telerik.Windows.Controls.ScheduleView
Imports Telerik.Windows.Controls
Imports System.ComponentModel
Imports System.Collections
Imports Telerik.Windows.Controls.ScheduleView.ICalendar

Public Class SqlAppointment
	Implements IAppointment, IExtendedAppointment, IObjectGenerator(Of IRecurrenceRule)

	Public Event RecurrenceRuleChanged As EventHandler
	Private _ExceptionOccurrences As List(Of SqlExceptionOccurrence)
	Private _ExceptionAppointments As List(Of SqlExceptionAppointment)
	Private _Resources As IList
	Private _TimeZone As TimeZoneInfo
	Private _RecurrenceRule As IRecurrenceRule

	Private Property Category_ As ICategory Implements IExtendedAppointment.Category
		Get
			Return TryCast(Me.Category, ICategory)
		End Get
		Set(value As ICategory)
			Me.Category = TryCast(value, Category)
		End Set
	End Property

	Private Property Importance_ As Importance Implements IExtendedAppointment.Importance
		Get
			Return Me.Importance
		End Get
		Set(value As Importance)
            Me.Importance = value
		End Set
	End Property

	Private Property TimeMarker_ As ITimeMarker Implements IExtendedAppointment.TimeMarker
		Get
			Return TryCast(Me.TimeMarker, ITimeMarker)
		End Get
		Set(value As ITimeMarker)
			Me.TimeMarker = TryCast(value, TimeMarker)
		End Set
	End Property

	Public Property TimeZone As TimeZoneInfo
		Get
			If Me._TimeZone Is Nothing Then
				Return TimeZoneInfo.Local
			End If

			Return Me._TimeZone
		End Get

		Set(value As TimeZoneInfo)
			If Not Object.Equals(Me._TimeZone, value) Then
				Me._TimeZone = value
				Me.OnPropertyChanged("TimeZone")
			End If
		End Set
	End Property

	Public Property RecurrenceRule As Telerik.Windows.Controls.ScheduleView.IRecurrenceRule Implements Telerik.Windows.Controls.ScheduleView.IAppointment.RecurrenceRule
		Get
			If Me._RecurrenceRule Is Nothing AndAlso Me.EntityState = System.Data.EntityState.Unchanged Then
				Me._RecurrenceRule = Me.GetRecurrenceRule(Me.RecurrencePattern)
			End If

			Return Me._RecurrenceRule
		End Get

		Set(value As Telerik.Windows.Controls.ScheduleView.IRecurrenceRule)
			If Not Object.Equals(Me.recurrenceRule, value) Then
				If value Is Nothing Then
					Me.RecurrencePattern = Nothing
				End If
				Me._RecurrenceRule = value
				Me.OnPropertyChanged("RecurrenceRule")
			End If
		End Set
	End Property

	Public ReadOnly Property Resources() As IList Implements IAppointment.Resources
		Get
			If Me._Resources Is Nothing Then
				Me._Resources = Me.SqlAppointmentResources.[Select](Function(ar) ar.SqlResource).ToList()
			End If

			Return Me._Resources
		End Get
	End Property

	Public Property Subject_ As String Implements Telerik.Windows.Controls.ScheduleView.IAppointment.Subject
		Get
			Return Me.Subject
		End Get
		Set(value As String)
			Me.Subject = value
		End Set
	End Property

	Public Property TimeZone_ As System.TimeZoneInfo Implements Telerik.Windows.Controls.ScheduleView.IAppointment.TimeZone
		Get
			Return Me.TimeZone
		End Get
		Set(value As System.TimeZoneInfo)
			Me.TimeZone = value
		End Set
	End Property

	Public Property End_ As Date Implements Telerik.Windows.Controls.ScheduleView.IAppointment.End
		Get
			Return Me.End
		End Get
		Set(value As Date)
			Me.End = value
		End Set
	End Property

	Public Property Start_ As Date Implements Telerik.Windows.Controls.ScheduleView.IAppointment.Start
		Get
			Return Me.Start
		End Get
		Set(value As Date)
			Me.Start = value
		End Set
	End Property

	Public ReadOnly Property End1 As Date Implements Telerik.Windows.Controls.ScheduleView.IDateSpan.End
		Get
			Return Me.End
		End Get
	End Property

	Public ReadOnly Property Start1 As Date Implements Telerik.Windows.Controls.ScheduleView.IDateSpan.Start
		Get
			Return Me.Start
		End Get
	End Property

	Public Property IsAllDayEvent_ As Boolean Implements Telerik.Windows.Controls.ScheduleView.IAppointment.IsAllDayEvent
		Get
			Return Me.IsAllDayEvent
		End Get
		Set(value As Boolean)
			Me.IsAllDayEvent = value
		End Set
	End Property

	Public Function Copy() As IAppointment
		Dim appointment As IAppointment = New SqlAppointment()
		appointment.CopyFrom(Me)
		Return appointment
	End Function

	Private Sub BeginEdit_() Implements System.ComponentModel.IEditableObject.BeginEdit
		If Me._ExceptionOccurrences Is Nothing Then
			Me._ExceptionOccurrences = New List(Of SqlExceptionOccurrence)()
		End If

		If Me._ExceptionAppointments Is Nothing Then
			Me._ExceptionAppointments = New List(Of SqlExceptionAppointment)()
		End If

		Me._ExceptionOccurrences.Clear()
		Me._ExceptionOccurrences.AddRange(Me.SqlExceptionOccurrences.ToList())

		Me._ExceptionAppointments.Clear()
		Me._ExceptionAppointments.AddRange(Me.SqlExceptionOccurrences.[Select](Function(o) o.Appointment).Where(Function(a) a IsNot Nothing).ToList())
	End Sub

	Private Sub CancelEdit_() Implements System.ComponentModel.IEditableObject.CancelEdit
        Dim exceptionOccurenceToRemove As IEnumerable(Of SqlExceptionOccurrence) = Me.SqlExceptionOccurrences.Except(Me._ExceptionOccurrences).ToList
		For Each ex As SqlExceptionOccurrence In exceptionOccurenceToRemove
			ScheduleViewRepository.Context.SqlExceptionOccurrences.DeleteObject(ex)
			If ex.Appointment IsNot Nothing Then
				ScheduleViewRepository.Context.SqlExceptionAppointments.DeleteObject(DirectCast(ex.Appointment, SqlExceptionAppointment))
				For Each resource In TryCast(ex.Appointment, SqlExceptionAppointment).SqlExceptionResources
					ScheduleViewRepository.Context.SqlExceptionResources.DeleteObject(resource)
				Next
			End If
		Next
	End Sub

	Private Sub EndEdit() Implements IEditableObject.EndEdit
		Dim resourceList = Me.SqlAppointmentResources.ToList()
		For Each resource In resourceList
			ScheduleViewRepository.Context.SqlAppointmentResources.DeleteObject(resource)
		Next

		For Each sqlResource In Me.Resources.OfType(Of SqlResource)()
			Me.SqlAppointmentResources.Add(New SqlAppointmentResource() With { _
			 .SqlAppointment = Me, _
			 .SqlResources_SqlResourceId = sqlResource.SqlResourceId _
			})
		Next

		Dim removedExceptionAppointments = Me._ExceptionAppointments.Except(Me.SqlExceptionOccurrences.[Select](Function(o) o.Appointment).OfType(Of SqlExceptionAppointment)())
		For Each exceptionAppointment In removedExceptionAppointments
			For Each item As SqlExceptionResource In exceptionAppointment.SqlExceptionResources
				ScheduleViewRepository.Context.SqlExceptionResources.DeleteObject(item)
			Next
		Next
	End Sub

	Public Function Equals_(other As Telerik.Windows.Controls.ScheduleView.IAppointment) As Boolean Implements System.IEquatable(Of Telerik.Windows.Controls.ScheduleView.IAppointment).Equals
		Dim otherAppointment = TryCast(other, SqlAppointment)
		Return otherAppointment IsNot Nothing AndAlso
		 other.Start = Me.Start AndAlso
		 other.[End] = Me.[End] AndAlso
		 other.Subject = Me.Subject AndAlso
		 Object.Equals(Me.CategoryID, otherAppointment.CategoryID) AndAlso
		 Object.Equals(Me.TimeMarker, otherAppointment.TimeMarker) AndAlso
		 Object.Equals(Me.TimeZone, otherAppointment.TimeZone) AndAlso
		 Me.IsAllDayEvent = other.IsAllDayEvent AndAlso
		 Object.Equals(Me.RecurrenceRule, other.RecurrenceRule)
	End Function

	Public Function CreateNew() As Telerik.Windows.Controls.ScheduleView.IRecurrenceRule Implements Telerik.Windows.Controls.IObjectGenerator(Of Telerik.Windows.Controls.ScheduleView.IRecurrenceRule).CreateNew
		Return Me.CreateDefaultRecurrenceRule()
	End Function

	Public Function CreateNew(item As Telerik.Windows.Controls.ScheduleView.IRecurrenceRule) As Telerik.Windows.Controls.ScheduleView.IRecurrenceRule Implements Telerik.Windows.Controls.IObjectGenerator(Of Telerik.Windows.Controls.ScheduleView.IRecurrenceRule).CreateNew
		Dim sqlRecurrenceRule = Me.CreateNew()
		sqlRecurrenceRule.CopyFrom(item)
		Return sqlRecurrenceRule
	End Function

	Public Function ShallowCopy() As IAppointment
		Dim appointment As IAppointment = New SqlExceptionAppointment()
		appointment.CopyFrom(Me)
		Return appointment
	End Function

	Private Sub CopyFrom(other As IAppointment) Implements ICopyable(Of IAppointment).CopyFrom
		Me.IsAllDayEvent = other.IsAllDayEvent
		Me.Start = other.Start
		Me.[End] = other.[End]
		Me.Subject = other.Subject

		Dim otherAppointment = TryCast(other, SqlAppointment)
		If otherAppointment Is Nothing Then
			Return
		End If

		Me.CategoryID = otherAppointment.CategoryID
		Me.TimeMarker = otherAppointment.TimeMarker
		Me.RecurrenceRule = If(other.RecurrenceRule Is Nothing, Nothing, TryCast(other.RecurrenceRule.Copy(), SqlRecurrenceRule))
		Me.RecurrencePattern = otherAppointment.RecurrencePattern

		Me.Resources.Clear()
		Me.Resources.AddRange(otherAppointment.Resources)

		Me.Body = otherAppointment.Body
	End Sub

	Private Function GetRecurrenceRule(pattern As String) As IRecurrenceRule
		If String.IsNullOrEmpty(pattern) Then
			Return Nothing
		End If

		Dim recurrenceRuleGenerator = TryCast(Me, IObjectGenerator(Of IRecurrenceRule))
		Dim recurrenceRule = If(recurrenceRuleGenerator IsNot Nothing, recurrenceRuleGenerator.CreateNew(), Me.CreateDefaultRecurrenceRule())
		Dim recurrencePattern = New RecurrencePattern()
		RecurrencePatternHelper.TryParseRecurrencePattern(pattern, recurrencePattern)
		recurrenceRule.Pattern = recurrencePattern
		For Each exception As SqlExceptionOccurrence In Me.SqlExceptionOccurrences
			recurrenceRule.Exceptions.Add(exception)
		Next

		Return recurrenceRule
	End Function

	Private Function CreateDefaultRecurrenceRule() As IRecurrenceRule
		Return New SqlRecurrenceRule(Me)
	End Function

	Public Function Copy_() As Telerik.Windows.Controls.ScheduleView.IAppointment Implements Telerik.Windows.Controls.ICopyable(Of Telerik.Windows.Controls.ScheduleView.IAppointment).Copy
		Dim appointment As IAppointment = New SqlAppointment()
		appointment.CopyFrom(Me)
		Return appointment
	End Function

	Public Event RecurrenceRuleChanged_(sender As Object, e As System.EventArgs) Implements Telerik.Windows.Controls.ScheduleView.IAppointment.RecurrenceRuleChanged
End Class

File: ScheduleView/Database/WPF_CS/Models/SqlAppointment.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Linq;
using Telerik.Windows.Controls;
using Telerik.Windows.Controls.ScheduleView;
using Telerik.Windows.Controls.ScheduleView.ICalendar;

namespace ScheduleViewDB
{
	public partial class SqlAppointment : IAppointment, IExtendedAppointment, IObjectGenerator<IRecurrenceRule>
	{
		public event EventHandler RecurrenceRuleChanged;
		private List<SqlExceptionOccurrence> exceptionOccurrences;
		private List<SqlExceptionAppointment> exceptionAppointments;
		private IList resources;
		private TimeZoneInfo timeZone;
		private IRecurrenceRule recurrenceRule;

		ITimeMarker IExtendedAppointment.TimeMarker
		{
			get
			{
				return this.TimeMarker as ITimeMarker;
			}
			set
			{
				this.TimeMarker = value as TimeMarker;
			}
		}

		ICategory IExtendedAppointment.Category
		{
			get
			{
				return this.Category as ICategory;
			}
			set
			{
				this.Category = value as Category;
			}
		}

		Importance IExtendedAppointment.Importance
		{
			get
			{
				return (Importance)this.Importance;
			}
			set
			{
				this.Importance = (int)value;
			}
		}

		public TimeZoneInfo TimeZone
		{
			get
			{
				if (this.timeZone == null)
				{
					return TimeZoneInfo.Local;
				}

				return this.timeZone;
			}

			set
			{
				if (this.timeZone != value)
				{
					this.timeZone = value;
					this.OnPropertyChanged("TimeZone");
				}
			}
		}

		public IList Resources
		{
			get
			{
				if (this.resources == null)
				{
					this.resources = this.SqlAppointmentResources.Select(ar => ar.SqlResource).ToList();
				}

				return this.resources;
			}
		}

		public IRecurrenceRule RecurrenceRule
		{
			get
			{
				if (this.recurrenceRule == null && this.EntityState == EntityState.Unchanged)
				{
					this.recurrenceRule = this.GetRecurrenceRule(this.RecurrencePattern);
				}

				return this.recurrenceRule;
			}

			set
			{
				if (this.recurrenceRule != value)
				{
					if (value == null)
					{
						this.RecurrencePattern = null;
					}
					this.recurrenceRule = value;
					this.OnPropertyChanged("RecurrenceRule");
				}
			}
		}

		public IAppointment Copy()
		{
			IAppointment appointment = new SqlAppointment();
			appointment.CopyFrom(this);
			return appointment;
		}

		void IEditableObject.BeginEdit()
		{
			if (this.exceptionOccurrences == null)
			{
				this.exceptionOccurrences = new List<SqlExceptionOccurrence>();
			}

			if (this.exceptionAppointments == null)
			{
				this.exceptionAppointments = new List<SqlExceptionAppointment>();
			}

			this.exceptionOccurrences.Clear();
			this.exceptionOccurrences.AddRange(this.SqlExceptionOccurrences.ToList());

			this.exceptionAppointments.Clear();
			this.exceptionAppointments.AddRange(this.SqlExceptionOccurrences.Select(o => o.Appointment).Where(a => a != null).ToList());
		}

		void IEditableObject.CancelEdit()
		{
			var exceptionOccurencesToRemove = this.SqlExceptionOccurrences.Except(this.exceptionOccurrences).ToList();
			foreach (var ex in exceptionOccurencesToRemove)
			{
				ScheduleViewRepository.Context.SqlExceptionOccurrences.DeleteObject(ex);
				if (ex.Appointment != null)
				{
					ScheduleViewRepository.Context.SqlExceptionAppointments.DeleteObject((SqlExceptionAppointment)ex.Appointment);
					foreach (var resource in (ex.Appointment as SqlExceptionAppointment).SqlExceptionResources)
					{
						ScheduleViewRepository.Context.SqlExceptionResources.DeleteObject(resource);
					}
				}
			}
		}

		void IEditableObject.EndEdit()
		{
			var temp = this.SqlAppointmentResources.ToList();
            var resources = this.Resources.OfType<SqlResource>().ToList();

			foreach (var item in temp)
			{
				ScheduleViewRepository.Context.SqlAppointmentResources.DeleteObject(item);
			}

            foreach (var sqlResource in resources)
			{
				ScheduleViewRepository.Context.AddToSqlAppointmentResources(new SqlAppointmentResource { SqlAppointment = this, SqlResources_SqlResourceId = sqlResource.SqlResourceId });
			}

			var removedExceptionAppointments = this.exceptionAppointments.Except(this.SqlExceptionOccurrences.Select(o => o.Appointment).OfType<SqlExceptionAppointment>());
			foreach (var exceptionAppointment in removedExceptionAppointments)
			{
				var excResources = exceptionAppointment.SqlExceptionResources.ToList();
				foreach (var item in excResources)
				{
					ScheduleViewRepository.Context.SqlExceptionResources.DeleteObject(item);
				}
			}
		}

		public bool Equals(IAppointment other)
		{
			var otherAppointment = other as SqlAppointment;
			return otherAppointment != null &&
				other.Start == this.Start &&
				other.End == this.End &&
				other.Subject == this.Subject &&
				this.CategoryID == otherAppointment.CategoryID &&
				this.TimeMarker == otherAppointment.TimeMarker &&
				this.TimeZone == otherAppointment.TimeZone &&
				this.IsAllDayEvent == other.IsAllDayEvent &&
				this.RecurrenceRule == other.RecurrenceRule;
		}

		public IRecurrenceRule CreateNew()
		{
			return this.CreateDefaultRecurrenceRule();
		}

		public IRecurrenceRule CreateNew(IRecurrenceRule item)
		{
			var sqlRecurrenceRule = this.CreateNew();
			sqlRecurrenceRule.CopyFrom(item);
			return sqlRecurrenceRule;
		}

		public IAppointment ShallowCopy()
		{
			var appointment = new SqlExceptionAppointment();
			appointment.CopyFrom(this);
			return appointment;
		}

		void ICopyable<IAppointment>.CopyFrom(IAppointment other)
		{
			this.IsAllDayEvent = other.IsAllDayEvent;
			this.Start = other.Start;
			this.End = other.End;
			this.Subject = other.Subject;

			var otherAppointment = other as SqlAppointment;
			if (otherAppointment == null)
				return;

			this.CategoryID = otherAppointment.CategoryID;
			this.TimeMarker = otherAppointment.TimeMarker;
			this.RecurrenceRule = other.RecurrenceRule == null ? null : other.RecurrenceRule.Copy() as SqlRecurrenceRule;
			this.RecurrencePattern = otherAppointment.RecurrencePattern;

			this.Resources.Clear();
			this.Resources.AddRange(otherAppointment.Resources);

			this.Body = otherAppointment.Body;
		}

		private IRecurrenceRule GetRecurrenceRule(string pattern)
		{
			if (string.IsNullOrEmpty(pattern))
			{
				return null;
			}

			var recurrenceRuleGenerator = this as IObjectGenerator<IRecurrenceRule>;
			var recurrenceRule = recurrenceRuleGenerator != null ? recurrenceRuleGenerator.CreateNew() : this.CreateDefaultRecurrenceRule();
			var recurrencePattern = new RecurrencePattern();
			RecurrencePatternHelper.TryParseRecurrencePattern(pattern, out recurrencePattern);
			recurrenceRule.Pattern = recurrencePattern;
			foreach (SqlExceptionOccurrence exception in this.SqlExceptionOccurrences)
			{
				recurrenceRule.Exceptions.Add(exception);
			}

			return recurrenceRule;
		}

		private IRecurrenceRule CreateDefaultRecurrenceRule()
		{
			return new SqlRecurrenceRule(this);
		}
	}
}
